src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs

```diff
@@ -643,7 +660,163 @@ private static size_t CurlSendCallback(IntPtr buffer, size_t size, size_t nitems
+                        }
+
+                        // Now Offset and Count are both accurate.  Determine how much data we can copy to libcurl...
+                        int availableData = sts._count - sts._offset;
+                        Debug.Assert(availableData > 0, "There must be some data still available.");
+
+                        // ... and copy as much of that as libcurl will allow.
+                        int bytesToCopy = Math.Min(availableData, length);
+                        Marshal.Copy(sts._buffer, sts._offset, buffer, bytesToCopy);
+                        multi.VerboseTrace("Copied " + bytesRead + " bytes from request stream", easy: easy);
+
+                        // Update the offset.  If we've gone through all of the data, reset the state 
+                        // so that the next time we're called back we'll do a new read.
+                        sts._offset += bytesToCopy;
+                        Debug.Assert(sts._offset <= sts._count, "Offset should never exceed count");
+                        if (sts._offset == sts._count)
+                        {
+                            sts.Update(null, 0, 0);
+                        }
+
+                        // Return the amount of data copied
+                        Debug.Assert(bytesToCopy > 0, "We should never return 0 bytes here.");
+                        return (size_t)bytesToCopy;
+                    }
+
+                    // sts was non-null but sts.Task was null, meaning there was no previous task/data
+                    // from which to satisfy any of this request.
+                }
+                else // sts == null
+                {
+                    // Allocate a transfer state object to use for the remainder of this request.
+                    easy._sendTransferState = sts = new EasyRequest.SendTransferState();
+                }
+
+                Debug.Assert(sts != null, "By this point we should have a transfer object");
+                Debug.Assert(sts._task == null, "There shouldn't be a task now.");
+
+                // If we get here, there was no previously read data available to copy.
+                // Initiate a new asynchronous read.
+                Task<int> asyncRead = easy._requestContentStream.ReadAsync(
+                    sts._buffer, 0, Math.Min(sts._buffer.Length, length), easy._cancellationToken);
+                Debug.Assert(asyncRead != null, "Badly implemented stream returned a null task from ReadAsync");
+
+                // Even though it's "Async", it's possible this read could complete synchronously or extremely quickly.  
+                // Check to see if it did, in which case we can also satisfy the libcurl request synchronously in this callback.
+                if (asyncRead.IsCompleted)
+                {
+                    multi.VerboseTrace("ReadAsync completed immediately", easy: easy);
+
+                    // Get the amount of data read.
```

Comment:
Nit: I see lot of code that is common to the if(sts!=null & sts.Task!=null) case. Consider separating out into a helper method